# Лабораторная работа №5: Поиск экстремума функции одной переменной с помощью ГА с бинарным представлением

**Дисциплина:** Генетические и эволюционные вычисления · **Вариант 3**

**Задача:** найти максимум $f(t) = (t - 1{,}3)\,\sin(0{,}5\pi t + 1)$ на отрезке $t \in [0;\ 7]$ простым (классическим) генетическим алгоритмом: бинарное кодирование, пропорциональная селекция (рулетка), одноточечный кроссинговер, побитовая мутация.

**Цель работы:** изучить основные понятия теории ГА, реализовать классический простой ГА с бинарным представлением и исследовать влияние его параметров на качество и надёжность поиска.

Ноутбук рассчитан на перезапуск по ячейкам в Docker-окружении репозитория (`make up`, http://localhost:8888). Полный прогон — около минуты.

In [1]:
import time
from pathlib import Path

import numpy as np
import pandas as pd
import plotly
import plotly.graph_objects as go
import plotly.io as pio
from plotly.subplots import make_subplots

OUTPUTS = Path('report/outputs')
FIGURES = Path('report/figures')
TABLES = Path('report/tables')
for p in (OUTPUTS, FIGURES, TABLES):
    p.mkdir(parents=True, exist_ok=True)

# Палитра проверена на различимость при дальтонизме
PALETTE = ['#2a78d6', '#eb6834', '#1baf7a', '#eda100', '#e87ba4', '#008300', '#4a3aa7', '#e34948']
BLUE_RAMP = ['#cde2fb', '#9ec5f4', '#6da7ec', '#3987e5', '#256abf', '#184f95', '#0d366b']
INK, INK2, MUTED, GRID, AXIS, SURFACE = '#0b0b0b', '#52514e', '#898781', '#e1e0d9', '#c3c2b7', '#fcfcfb'
GOAL = '#d03b3b'

axis_style = dict(gridcolor=GRID, linecolor=AXIS, zeroline=False, ticks='outside', tickcolor=AXIS, ticklen=4,
                  title_font=dict(size=14, color=INK2), tickfont=dict(size=12, color=INK2))
pio.templates['lab'] = go.layout.Template(layout=dict(
    font=dict(family='system-ui, -apple-system, "Segoe UI", Roboto, Arial, sans-serif', size=14, color=INK),
    paper_bgcolor=SURFACE, plot_bgcolor=SURFACE, colorway=PALETTE,
    xaxis=axis_style, yaxis=axis_style,
    title=dict(x=0.01, xanchor='left', font=dict(size=18, color=INK)),
    legend=dict(bgcolor='rgba(252,252,251,0.85)', bordercolor=GRID, borderwidth=1, font=dict(size=12)),
    hoverlabel=dict(bgcolor='white', bordercolor=AXIS, font=dict(size=12, color=INK)),
    margin=dict(l=70, r=30, t=80, b=60),
))
pio.templates.default = 'plotly_white+lab'
PLOT_CONFIG = dict(displaylogo=False, responsive=True, toImageButtonOptions=dict(format='png', scale=2))


def save(fig, name, width=1000, height=520, png=True, png_fig=None):
    """HTML (интерактивный) в report/outputs и PNG (для отчёта) в report/figures."""
    fig.write_html(OUTPUTS / f'{name}.html', include_plotlyjs='cdn', config=PLOT_CONFIG, auto_play=False)
    if png:
        (png_fig or fig).write_image(FIGURES / f'{name}.png', width=width, height=height, scale=2)


def to_latex(df, name, caption, label, col_format=None):
    """Таблица результатов для отчёта: числа в отчёте совпадают с ноутбуком."""
    body = df.to_latex(index=False, escape=False, column_format=col_format or 'c' * len(df.columns))
    tex = ('\\begin{table}[H]\n\\caption{' + caption + '}\n\\label{' + label + '}\n\\centering\n\\small\n'
           + body + '\\end{table}\n')
    (TABLES / f'{name}.tex').write_text(tex, encoding='utf-8')


def anim_controls(steps, prefix='поколение: ', duration=250):
    """Кнопки «Старт / Пауза / Сброс» и слайдер для анимации Plotly."""
    play = dict(frame=dict(duration=duration, redraw=False), transition=dict(duration=int(duration * 0.8)),
                fromcurrent=True, mode='immediate')
    still = dict(frame=dict(duration=0, redraw=False), mode='immediate', transition=dict(duration=0))
    return dict(
        updatemenus=[dict(type='buttons', direction='left', x=1.0, xanchor='right', y=1.1, yanchor='bottom',
                          showactive=False, buttons=[
                              dict(label='▶ Старт', method='animate', args=[None, play]),
                              dict(label='⏸ Пауза', method='animate', args=[[None], still]),
                              dict(label='⟲ Сброс', method='animate', args=[[str(steps[0])], still])])],
        sliders=[dict(active=0, x=0, len=1, y=-0.08, yanchor='top', pad=dict(t=30),
                      currentvalue=dict(prefix=prefix, font=dict(size=13, color=INK2)),
                      steps=[dict(label=str(s), method='animate', args=[[str(s)], still]) for s in steps])])


print('numpy', np.__version__, '| pandas', pd.__version__, '| plotly', plotly.__version__)
from scipy.optimize import minimize_scalar

numpy 2.4.4 | pandas 3.0.2 | plotly 6.7.0


## 1. Целевая функция и эталонный максимум

Эталон для оценки ГА находим численно: перебор по сетке из 700 001 точки и уточнение методом Брента.

In [2]:
A, B = 0.0, 7.0


def f(t):
    return (t - 1.3) * np.sin(0.5 * np.pi * t + 1)


t_dense = np.linspace(A, B, 700_001)
i0 = int(np.argmax(f(t_dense)))
opt = minimize_scalar(lambda x: -f(x), bounds=(t_dense[i0] - 1e-3, t_dense[i0] + 1e-3), method='bounded',
                      options=dict(xatol=1e-12))
T_STAR, F_STAR = float(opt.x), float(-opt.fun)
d = np.diff(np.sign(np.diff(f(t_dense))))
loc_max = t_dense[np.where(d < 0)[0] + 1]
print(f'Глобальный максимум: t* = {T_STAR:.6f}, f(t*) = {F_STAR:.6f}')
print('Локальные максимумы:', ', '.join(f't = {x:.3f} (f = {f(x):.3f})' for x in loc_max))
print(f'Значения на концах: f(0) = {f(A):.3f}, f(7) = {f(B):.3f}')

t_plot = np.linspace(A, B, 1000)
fig = go.Figure()
fig.add_trace(go.Scatter(x=t_plot, y=f(t_plot), mode='lines', name='f(t)', line=dict(color=PALETTE[0], width=2.5),
                         hovertemplate='t = %{x:.3f}<br>f = %{y:.4f}<extra></extra>'))
fig.add_trace(go.Scatter(x=[T_STAR], y=[F_STAR], mode='markers+text', name='глобальный максимум',
                         marker=dict(symbol='star', size=16, color=PALETTE[1], line=dict(color=SURFACE, width=1.5)),
                         text=[f'  t* = {T_STAR:.4f}, f* = {F_STAR:.4f}'], textposition='middle right',
                         textfont=dict(color=INK2)))
fig.update_layout(title='Вариант 3: f(t) = (t − 1.3)·sin(0.5πt + 1), t ∈ [0; 7]', xaxis_title='t', yaxis_title='f(t)',
                  legend=dict(x=0.01, y=0.02, yanchor='bottom'), height=460)
save(fig, '01_function', height=460)
fig.show()

Глобальный максимум: t* = 4.488826, f(t*) = 3.127117
Локальные максимумы: t = 1.332 (f = 0.002), t = 4.489 (f = 3.127)
Значения на концах: f(0) = -1.094, f(7) = -3.080


## 2. Бинарное кодирование

Число $t \in [a;\ b]$ кодируется строкой из $L$ бит: $t = a + \dfrac{\mathrm{dec}(s)}{2^L - 1}(b - a)$. Шаг сетки представимых значений (точность) $\Delta = (b-a)/(2^L-1)$ — длину хромосомы выбирают по требуемой точности.

In [3]:
def decode(pop, a=A, b=B):
    """Декодирование популяции (N, L) бинарных строк в значения t."""
    L = pop.shape[1]
    return a + (pop @ (2 ** np.arange(L - 1, -1, -1))) * (b - a) / (2 ** L - 1)


def encode(t, L, a=A, b=B):
    v = int(round((t - a) / (b - a) * (2 ** L - 1)))
    return np.array(list(format(v, f'0{L}b')), dtype=np.int8)


s = encode(T_STAR, 20)
print(f't* = {T_STAR:.6f} → {"".join(map(str, s))} → {decode(s[None])[0]:.6f}')
df_prec = pd.DataFrame([dict(L=L, **{'шаг Δ': (B - A) / (2 ** L - 1),
                                      'лучшее представимое t': decode(encode(T_STAR, L)[None])[0]})
                        for L in (8, 10, 16, 20, 24)])
df_prec['потеря по f'] = F_STAR - f(df_prec['лучшее представимое t'])
df_prec.style.format({'шаг Δ': '{:.2e}', 'лучшее представимое t': '{:.6f}', 'потеря по f': '{:.1e}'}).hide(axis='index')

t* = 4.488826 → 10100100001010011010 → 4.488825


L,шаг Δ,лучшее представимое t,потеря по f
8,2.75e-02,4.501961,7.2e-04
10,6.84e-03,4.488759,1.9e-08
16,1.07e-04,4.488823,5.0e-11
20,6.68e-06,4.488825,3.6e-12
24,4.17e-07,4.488826,1.6e-14


## 3. Простой генетический алгоритм

- **Инициализация:** $N$ случайных бинарных строк.
- **Селекция — рулетка:** $p_i = f'_i / \sum f'_j$, где $f'_i = f_i - \min f + \varepsilon$ (сдвиг, так как $f$ принимает отрицательные значения).
- **Одноточечный кроссинговер** с вероятностью $P_c$: пары обмениваются «хвостами» после случайной точки разреза.
- **Побитовая мутация:** каждый бит инвертируется с вероятностью $P_m$.
- **Элитизм:** лучшая найденная особь замещает худшего потомка (можно отключить для сравнения).

Для каждого запуска фиксируется поколение, на котором впервые найдено решение с $|t - t^*| < 10^{-3}$ («попадание»).

In [4]:
TOL = 1e-3
DEFAULT = dict(pop_size=60, L=20, p_c=0.85, p_m=0.01, n_gen=200, elitism=True)


def run_ga(pop_size=60, L=20, p_c=0.85, p_m=0.01, n_gen=200, elitism=True, seed=0, track=False):
    rng = np.random.default_rng(seed)
    pop = rng.integers(0, 2, (pop_size, L), dtype=np.int8)
    t = decode(pop); fit = f(t)
    best_i = int(np.argmax(fit)); best_chrom, best_fit = pop[best_i].copy(), fit[best_i]
    hist_best, hist_mean, snaps, hit_gen = [best_fit], [fit.mean()], [t.copy()] if track else None, None
    if abs(decode(best_chrom[None])[0] - T_STAR) < TOL:
        hit_gen = 0
    for g in range(1, n_gen + 1):
        # рулетка
        w = fit - fit.min() + 1e-9
        pop = pop[rng.choice(pop_size, pop_size, p=w / w.sum())]
        # одноточечный кроссинговер (векторно по парам)
        n_pairs = pop_size // 2
        do = rng.random(n_pairs) < p_c
        cut = rng.integers(1, L, n_pairs)
        mask = (np.arange(L)[None, :] >= cut[:, None]) & do[:, None]
        a, b = pop[0:2 * n_pairs:2].copy(), pop[1:2 * n_pairs:2].copy()
        pop[0:2 * n_pairs:2] = np.where(mask, b, a)
        pop[1:2 * n_pairs:2] = np.where(mask, a, b)
        # побитовая мутация
        pop ^= (rng.random(pop.shape) < p_m).astype(np.int8)
        t = decode(pop); fit = f(t)
        if elitism:
            w_i = int(np.argmin(fit))
            if fit[w_i] < best_fit:
                pop[w_i], t[w_i], fit[w_i] = best_chrom, decode(best_chrom[None])[0], best_fit
        cur = int(np.argmax(fit))
        if fit[cur] > best_fit:
            best_fit, best_chrom = fit[cur], pop[cur].copy()
        if hit_gen is None and abs(decode(best_chrom[None])[0] - T_STAR) < TOL:
            hit_gen = g
        hist_best.append(fit.max()); hist_mean.append(fit.mean())
        if track:
            snaps.append(t.copy())
    best_t = decode(best_chrom[None])[0]
    return dict(best_t=best_t, best_f=best_fit, chrom=best_chrom, hist_best=np.array(hist_best),
                hist_mean=np.array(hist_mean), snaps=snaps, hit_gen=hit_gen, success=abs(best_t - T_STAR) < TOL)


res = run_ga(**DEFAULT, seed=0, track=True)
print(f'Найдено: t = {res["best_t"]:.6f}, f(t) = {res["best_f"]:.6f}; хромосома {"".join(map(str, res["chrom"]))}')
print(f'Отклонение от эталона: |Δt| = {abs(res["best_t"] - T_STAR):.2e}, Δf = {F_STAR - res["best_f"]:.2e}; '
      f'попадание в окрестность 1e-3 на поколении {res["hit_gen"]}')

Найдено: t = 4.488825, f(t) = 3.127117; хромосома 10100100001010011010
Отклонение от эталона: |Δt| = 9.24e-07, Δf = 3.60e-12; попадание в окрестность 1e-3 на поколении 7


In [5]:
# Динамика лучшего и среднего значения
g = np.arange(len(res['hist_best']))
fig = go.Figure()
fig.add_trace(go.Scatter(x=g, y=res['hist_best'], mode='lines', name='лучшая особь поколения',
                         line=dict(color=PALETTE[0], width=2.5)))
fig.add_trace(go.Scatter(x=g, y=res['hist_mean'], mode='lines', name='среднее по популяции',
                         line=dict(color=PALETTE[1], width=2)))
fig.add_hline(y=F_STAR, line=dict(color=GOAL, dash='dash', width=1.5), annotation_text=f'f* = {F_STAR:.4f}',
              annotation_position='bottom right', annotation_font_color=GOAL)
fig.update_layout(title='Сходимость простого ГА (N = 60, L = 20, Pc = 0.85, Pm = 0.01)', xaxis_title='поколение',
                  yaxis_title='f(t)', hovermode='x unified', legend=dict(x=0.99, xanchor='right', y=0.05, yanchor='bottom'),
                  height=440)
save(fig, '02_convergence', height=440)
fig.show()

In [6]:
# Анимация популяции на графике функции (каждые 5 поколений первые 60, затем каждые 20)
steps = list(range(0, 60, 5)) + list(range(60, 201, 20))
fig = go.Figure()
fig.add_trace(go.Scatter(x=t_plot, y=f(t_plot), mode='lines', name='f(t)', line=dict(color=INK2, width=2)))
t0 = res['snaps'][0]
fig.add_trace(go.Scatter(x=t0, y=f(t0), mode='markers', name='особи',
                         marker=dict(color=PALETTE[1], size=9, opacity=0.75, line=dict(color=SURFACE, width=1))))
fig.frames = [go.Frame(name=str(s), data=[go.Scatter(x=res['snaps'][s], y=f(res['snaps'][s]))], traces=[1],
                       layout=go.Layout(title_text=f'Популяция на графике функции — поколение {s}'))
              for s in steps]
fig.update_layout(title='Популяция на графике функции — поколение 0', xaxis_title='t', yaxis_title='f(t)',
                  height=560, margin=dict(t=110, b=120), legend=dict(x=0.01, y=0.02, yanchor='bottom'),
                  **anim_controls(steps))
png = make_subplots(rows=1, cols=3, shared_yaxes=True, horizontal_spacing=0.03,
                    subplot_titles=[f'поколение {s}' for s in (0, 10, 200)])
for c, s in enumerate((0, 10, 200), start=1):
    png.add_trace(go.Scatter(x=t_plot, y=f(t_plot), mode='lines', line=dict(color=INK2, width=2),
                             name='f(t)', showlegend=c == 1), 1, c)
    png.add_trace(go.Scatter(x=res['snaps'][s], y=f(res['snaps'][s]), mode='markers', name='особи', showlegend=c == 1,
                             marker=dict(color=PALETTE[1], size=8, opacity=0.75, line=dict(color=SURFACE, width=1))), 1, c)
    png.update_xaxes(title_text='t', row=1, col=c)
png.update_layout(title='Популяция простого ГА на графике функции', height=420,
                  legend=dict(orientation='h', x=0.5, xanchor='center', y=-0.22))
save(fig, '03_population', width=1300, height=420, png_fig=png)
fig.show()

## 4. Надёжность ГА: 50 независимых запусков

Один удачный запуск ничего не доказывает — ГА стохастичен. Оценим долю успешных запусков (найден глобальный максимум с точностью $10^{-3}$) и число поколений до попадания.

In [7]:
N_RUNS = 50
runs = [run_ga(**DEFAULT, seed=100 + s) for s in range(N_RUNS)]
succ = np.array([r['success'] for r in runs])
hits = np.array([r['hit_gen'] for r in runs if r['hit_gen'] is not None])
errs = np.array([abs(r['best_t'] - T_STAR) for r in runs])
print(f'Успешных запусков: {succ.sum()}/{N_RUNS}; поколений до попадания: медиана {np.median(hits):.0f}, '
      f'90-й перцентиль {np.percentile(hits, 90):.0f}, максимум {hits.max()}')
print(f'|t − t*| в успешных запусках: медиана {np.median(errs[succ]):.1e}, максимум {errs[succ].max():.1e}; '
      f'вычислений функции до попадания (медиана): {np.median(hits) * 60:.0f}')

Успешных запусков: 47/50; поколений до попадания: медиана 11, 90-й перцентиль 39, максимум 168
|t − t*| в успешных запусках: медиана 9.2e-07, максимум 9.2e-07; вычислений функции до попадания (медиана): 660


## 5. Влияние параметров (30 запусков на каждое значение)

Меняется один параметр, остальные — по умолчанию ($N = 60$, $L = 20$, $P_c = 0{,}85$, $P_m = 0{,}01$, 200 поколений). Для малых популяций и длинного поиска важна не только средняя, но и **худшая** точность, поэтому приводятся доля успехов и медиана поколений до попадания.

In [8]:
GRID = {'pop_size': [10, 20, 60, 100], 'L': [8, 12, 20, 24], 'p_c': [0.0, 0.5, 0.85, 0.95],
        'p_m': [0.0, 0.001, 0.01, 0.05, 0.2], 'n_gen': [10, 25, 50, 200]}
LABELS = {'pop_size': 'размер популяции N', 'L': 'длина хромосомы L', 'p_c': 'вероятность кроссинговера Pc',
          'p_m': 'вероятность мутации Pm', 'n_gen': 'число поколений'}
R = 30
rows = []
tic = time.perf_counter()
for par, vals in GRID.items():
    for v in vals:
        rr = [run_ga(**{**DEFAULT, par: v}, seed=1000 + s) for s in range(R)]
        hg = [r['hit_gen'] for r in rr if r['hit_gen'] is not None]
        rows.append(dict(параметр=LABELS[par], par=par, значение=v,
                         успехов=np.mean([r['success'] for r in rr]) * 100,
                         среднее_f=np.mean([r['best_f'] for r in rr]), худшее_f=np.min([r['best_f'] for r in rr]),
                         медиана_поколений=np.median(hg) if hg else np.nan))
df_par = pd.DataFrame(rows)
print(f'{len(rows) * R} запусков за {time.perf_counter() - tic:.1f} с')
df_par.drop(columns='par').style.format({'успехов': '{:.0f} %', 'среднее_f': '{:.4f}', 'худшее_f': '{:.4f}',
                                         'медиана_поколений': '{:.0f}'}).hide(axis='index')

630 запусков за 4.0 с


параметр,значение,успехов,среднее_f,худшее_f,медиана_поколений
размер популяции N,10.000000,27 %,3.1082,3.0745,60
размер популяции N,20.000000,47 %,3.1218,3.0745,16
размер популяции N,60.000000,97 %,3.1271,3.1270,15
размер популяции N,100.000000,100 %,3.1271,3.1271,12
длина хромосомы L,8.000000,0 %,3.1264,3.1264,nan
длина хромосомы L,12.000000,93 %,3.1253,3.0739,11
длина хромосомы L,20.000000,97 %,3.1271,3.1270,15
длина хромосомы L,24.000000,97 %,3.1254,3.0745,16
вероятность кроссинговера Pc,0.000000,90 %,3.1271,3.1270,24
вероятность кроссинговера Pc,0.500000,93 %,3.1271,3.1270,17


In [9]:
fig = make_subplots(rows=1, cols=5, shared_yaxes=True, horizontal_spacing=0.025,
                    subplot_titles=[LABELS[p] for p in GRID])
for c, par in enumerate(GRID, start=1):
    d = df_par[df_par.par == par]
    fig.add_trace(go.Bar(x=[f'{v:g}' for v in d.значение], y=d.успехов, marker=dict(color=PALETTE[0],
                         line=dict(color=SURFACE, width=2)), showlegend=False,
                         text=[f'{v:.0f}%' for v in d.успехов], textposition='outside', textfont=dict(size=11, color=INK2),
                         customdata=np.c_[d.худшее_f, d.медиана_поколений],
                         hovertemplate='%{x}: успехов %{y:.0f}%<br>худшее f = %{customdata[0]:.4f}<br>'
                                       'поколений до попадания: %{customdata[1]:.0f}<extra></extra>'), 1, c)
fig.update_yaxes(range=[0, 115], title_text='доля успешных запусков, %', row=1, col=1)
fig.update_xaxes(type='category')
fig.update_layout(title='Доля запусков, нашедших глобальный максимум (|t − t*| < 10⁻³), 30 запусков на значение',
                  height=430, bargap=0.3)
save(fig, '04_parameters', width=1400, height=430)
fig.show()

## 6. Роль элитизма

Сравним средние кривые лучшей особи текущего поколения с элитизмом и без (30 запусков, $P_m = 0{,}05$ — при заметной мутации без элитизма лучшая особь легко теряется).

In [10]:
curves = {}
for el in (True, False):
    rr = [run_ga(**{**DEFAULT, 'p_m': 0.05, 'elitism': el}, seed=2000 + s) for s in range(R)]
    curves[el] = (np.mean([r['hist_best'] for r in rr], axis=0), np.mean([r['success'] for r in rr]) * 100,
                  np.mean([r['hist_best'][-1] for r in rr]))
fig = go.Figure()
for k, (el, name) in enumerate([(True, 'с элитизмом'), (False, 'без элитизма')]):
    fig.add_trace(go.Scatter(x=np.arange(len(curves[el][0])), y=curves[el][0], mode='lines',
                             name=f'{name}: f последнего поколения {curves[el][2]:.4f}', line=dict(color=PALETTE[k], width=2.5)))
fig.add_hline(y=F_STAR, line=dict(color=GOAL, dash='dash', width=1.5))
fig.update_layout(title='Лучшая особь поколения (среднее по 30 запускам, Pm = 0.05)', xaxis_title='поколение',
                  yaxis_title='f(t)', yaxis_range=[2.9, 3.16], legend=dict(x=0.99, xanchor='right', y=0.05, yanchor='bottom'),
                  height=420)
save(fig, '05_elitism', height=420)
fig.show()
print({('с элитизмом' if k else 'без элитизма'): f'успехов {v[1]:.0f} %, f последнего поколения {v[2]:.5f}'
       for k, v in curves.items()})

{'с элитизмом': 'успехов 100 %, f последнего поколения 3.12712', 'без элитизма': 'успехов 100 %, f последнего поколения 3.12270'}


## 7. Таблицы для отчёта

In [11]:
t1 = pd.DataFrame({'Параметр': ['$t^*$ (эталон)', '$f(t^*)$ (эталон)', '$t$ (ГА)', '$f(t)$ (ГА)', '$|t - t^*|$',
                                'Поколение попадания', 'Успешных запусков из 50', 'Поколений до попадания (медиана / макс.)'],
                   'Значение': [f'{T_STAR:.6f}', f'{F_STAR:.6f}', f'{res["best_t"]:.6f}', f'{res["best_f"]:.6f}',
                                f'{abs(res["best_t"] - T_STAR):.1e}', str(res['hit_gen']), f'{succ.sum()}',
                                f'{np.median(hits):.0f} / {hits.max()}']})
to_latex(t1, 'tab_result', 'Результат поиска максимума (N = 60, L = 20, $P_c = 0{,}85$, $P_m = 0{,}01$)', 'tab:result', 'lc')

t2 = df_par.copy()
t2 = pd.DataFrame({'Параметр': t2.параметр.str.replace('вероятность кроссинговера Pc', 'вероятность $P_c$')
                   .str.replace('вероятность мутации Pm', 'вероятность $P_m$'),
                   'Значение': t2.значение.map(lambda v: f'{v:g}'),
                   'Успехов, \\%': t2.успехов.map(lambda v: f'{v:.0f}'), 'Среднее $f$': t2.среднее_f.map(lambda v: f'{v:.4f}'),
                   'Худшее $f$': t2.худшее_f.map(lambda v: f'{v:.4f}'),
                   'Поколений': t2.медиана_поколений.map(lambda v: '---' if np.isnan(v) else f'{v:.0f}')})
to_latex(t2, 'tab_params', 'Влияние параметров ГА (30 запусков на значение)', 'tab:params', 'lccccc')
print(sorted(p.name for p in TABLES.glob('*.tex')))

['tab_params.tex', 'tab_result.tex']


## Выводы

1. **Результат.** Простой ГА (N = 60, L = 20, Pc = 0,85, Pm = 0,01) находит глобальный максимум $t^* = 4{,}48883$, $f^* = 3{,}12712$ с погрешностью $|t - t^*| = 9 \cdot 10^{-7}$ — это точность сетки 20-битного кода ($6{,}7 \cdot 10^{-6}$). В окрестность $10^{-3}$ алгоритм попадает на 7-м поколении, т.е. после ~420 вычислений функции.
2. **Надёжность.** Из 50 независимых запусков 47 успешны, медиана — 11 поколений до попадания. Три «неудачи» — не ложный максимум, а остановка на плоской вершине: $f = 3{,}1270$ при $|t - t^*| > 10^{-3}$. Около вершины значения $f$ почти не различаются, рулетка теряет давление отбора, и тонкую доводку выполняет только мутация.
3. **Размер популяции** — главный фактор надёжности: при N = 10 успешны 27 % запусков, при N = 20 — 47 %, при N = 60 — 97 %, при N = 100 — 100 %.
4. **Длина хромосомы задаёт предельную точность:** при L = 8 шаг сетки 0,027 больше допуска $10^{-3}$, и попадание невозможно в принципе (0 %), хотя $f$ отличается от максимума лишь на $7 \cdot 10^{-4}$. Удлинение с 20 до 24 бит точность не улучшает: пространство поиска растёт в 16 раз.
5. **Мутация необходима:** без неё (Pm = 0) успешны только 50 % запусков — популяция вырождается и не может получить недостающие биты. При Pm = 0,05–0,2 успех 100 %. Кроссинговер ускоряет поиск (медиана 24 → 12 поколений при Pc 0 → 0,95), но на одномодальной по сути функции не критичен.
6. **Элитизм** гарантирует монотонный рост лучшего решения: при Pm = 0,05 без элитизма лучшая особь последнего поколения в среднем хуже ($f = 3{,}1227$ против $3{,}1271$), найденный максимум теряется.